### PR TITLE
More strict validation for srt files.

### PR DESCRIPTION
**Pull request by RamanYeda**
Improve validation of srt files.

### CODE REVIEW

**Comment by vkhramtsov** [comment]
@delphiki, any news about this PR?

File: tests/Captioning/ConverterTest.php [hunk 1]
```diff
@@ -16,23 +16,23 @@ public function testSubrip2WebvttConversion()
 
         $content = "WEBVTT
 
-00:00:00.000 --> 00:00:20.000
+00:00:01.000 --> 00:00:20.000
```
Comment (delphiki) [inline tests/Captioning/ConverterTest.php]: Why did you change all the timestamps exactly?

No that it really matters, but I was just wondering.

File: src/Captioning/Format/SubripFile.php [hunk 2]
```diff
@@ -116,4 +155,18 @@ private function validateOptions(array $options)
...
+     * @param \DateTime $endTimeline
+     * @return boolean
+     */
+    private function validateTimelines(\DateTime $startTimeline, \DateTime $endTimeline)
+    {
+        if ($startTimeline >= $endTimeline) {
+            return false;
+        }
+
+        return true;
```
Comment (delphiki) [inline src/Captioning/Format/SubripFile.php]:
```php
return ($startTimeline >= $endTimelin);
```
would be even better :)

File: src/Captioning/Format/SubripFile.php [hunk 3]
```diff
@@ -6,9 +6,28 @@
...
+        (?:
+            \1(?<=\r\n|\r|\n)[\d]+\1
+            [\d]{2}:[\d]{2}:[\d]{2},[\d]{3}
+            [ ]-->[ ]
+            [\d]{2}:[\d]{2}:[\d]{2},[\d]{3}
+            \1(?:[\S ]+\1)+
+        )*
+        \1?
+        $/xu'
+    ;
```
Comment (delphiki) [inline src/Captioning/Format/SubripFile.php]: Really nice edit here, clarifies the whole regex, nice job!

**Comment by delphiki** [comment]
Sorry for the late review (@vkhramtsov is used to it ;-)), but overall, great job @RamanYeda, really nice additions!

I'll just wait for your answer about some of my comments and I'll merge it. :+1:

File: src/Captioning/Format/SubripFile.php (hunk 2 above)
Comment (RamanYeda) [inline src/Captioning/Format/SubripFile.php]: Yes, sorry about this bulky code. Fixed it.

File: src/Captioning/Format/SubripFile.php (hunk 3 above)
Comment (RamanYeda) [inline src/Captioning/Format/SubripFile.php]: Thank you. I am often use /x modifier in regular expressions when they are too difficult for understanding in one line.

File: tests/Captioning/ConverterTest.php (hunk 1 above)
Comment (RamanYeda) [inline tests/Captioning/ConverterTest.php]: Accordingly to current logic parse method checks result of regular expression execution.
First thing what we need to do - check current subtitle order ($subtitle[0] != $subtitleOrder++).
If we okay with order number then second what we want - validate timelines. The imposition of timelines is unacceptalbe.
And there is the problem. We do timeline validation through PHP DateTime objects comparison. And (i don't know the reason of this and don't founded any words about this on php.net) DateTime comparison refuse to compare milliseconds. I mean it's slill equals:
    $a = \DateTime::createFromFormat('H:i:s,u', '00:00:00,000');
    $b = \DateTime::createFromFormat('H:i:s,u', '00:00:00,999');
    print ($a >= $b); // will print 1 because condition statement returns true.

Now in context of converter test: we start validation from 00:00:00.000 and on first iteration in foreach (in parse method) after validation subtitle order we do comparise 00:00:00.000 and 00:00:00.000 - so validateTimelines will return false and validation will fail. Same result will be if first subtitle timeline will stars from lwer than 1 seconds. Because 00:00:00.000 and 00:00:00.999 still equal for PHP DateTime comparison.

I think what this problem can be resolved in next improvement (potentially need rewrine validateTimelines method logic to more complex than just comparing two DateTime objects).

**Comment by RamanYeda** [comment]
Well, thank you. This is my first attempt to open source contribution and I trying to do my best.

File: tests/Captioning/ConverterTest.php (hunk 1 above)
Comment (delphiki) [inline tests/Captioning/ConverterTest.php]: I see your point, and understand the limitations, but that's also the reason why I can't accept this merge request in this state.

I can't merge code that will include regression. 
Lots of subtitles start at 00:00:00.000, I can't push this to the current release branch as it would break with those files' parsing.

This could be a lead on how to bypass the DateTime limitation: http://stackoverflow.com/questions/17089533/comparing-times-with-milliseconds-and-without-dates

File: src/Captioning/Format/SubripFile.php [hunk 4]
```diff
@@ -116,4 +155,14 @@ private function validateOptions(array $options)
...
     }
+
+    /**
+     * @param \DateTime $startTimeline
+     * @param \DateTime $endTimeline
+     * @return boolean
+     */
+    private function validateTimelines(\DateTime $startTimeline, \DateTime $endTimeline)
+    {
+        return !($startTimeline >= $endTimeline);
```
Comment (delphiki) [inline src/Captioning/Format/SubripFile.php]: or just simply (the negation isn't necessary):
```php
return ($startTimeline < $endTimeline);
```

But considering the fact that milliseconds can't be compared with DateTimes, this method will certainly change.

Comment (RamanYeda) [inline src/Captioning/Format/SubripFile.php]: Added milliseconds comparison if both DateTime objects are equal.

File: tests/Captioning/ConverterTest.php (hunk 1 above)
Comment (RamanYeda) [inline tests/Captioning/ConverterTest.php]: Thanks for link. It help me to solve the milliseconds problem.

**Comment by delphiki** [comment]
Great, thanks! :+1:
